File: Leetcode/top_interview_questions/easy/arrays/5.single_number.js
/**
 * @param {number[]} nums
 * @return {number}
 */

// This is the first approach that use extra space O(n)
var singleNumber = function (nums) {
  let temp = [];
  for (let i = 0; i < nums.length; i++) {
    if (!nums.slice(i + 1).includes(nums[i]) && !temp.includes(nums[i])) {
      return nums[i];
    }
    temp[i] = nums[i];
  }
};

// This is the approach based on XOR operation
var singleNumber = function (nums) {
  let temp = 0;
  for (let i = 0; i < nums.length; i++) {
    temp = temp ^ nums[i];
  }

  return temp;
};

console.log(singleNumber([2, 2, 1]));

File: Leetcode/collections/top_interview_questions/easy/arrays/5.single_number.py
class Solution:
    def singleNumber(self, nums: list[int]) -> int:
        sample = 23 ^ 12
        print("result is", sample)
        
        result = 0
        for x in nums:
           result = result ^ x
        return result
     
sample = Solution()
print(sample.singleNumber([4,1,2,1,2]))